### ISSUE TITLE
Towards a fix to the loss of subject reduction with corecursion?

### ISSUE THREAD

**Issue by herbelin** [coq/coq#6768]
I wonder whether we could solve the failure of subject reduction with corecursion by using an asymmetric conversion algorithm. Let me remind the failure of subject reduction with (so-called) co-fixpoints:
```coq
CoInductive I := C : I -> I.
CoFixpoint infty := C infty.
Definition unfold : infty = C infty :=
  match infty as x return match x with C n => x = C n end with
  | C n => eq_refl (C n)
  end.
Fail Definition nf_unfold : infty = C infty := Eval lazy in unfold.
```
It fails because we have to compare `C infty = C infty ≤ infty = C infty`, i.e. `C infty ≤ infty` and reduction does not allow to unfold `infty` since noone is asking for destructing it.

But, to some extend, someone is asking for destructing it: it is the other side of the conversion problem. So, the same way as we implemented η where the presence of a constructor on one side makes legitimate to η-expand on the other side, can't we accept that a constructor on one side makes legitimate to cofix-expand on the other side.

The problem is, that, on the contrary of η-expansion for a connective, there is no finite type structure to ensure termination of the process, in the sense that η-expanding on one side might lead to require to η-expand on the other side, and this infinitely many often.

So, the question is: could we solve the problem by allowing cofix-expansion only on one side? After all, subject reduction generates oriented conversion problems (i.e. "type of subject" ≤ "object"), so we may be able to preserve this information of orientation in the conversion algorithm.

What do you think?

**Comment by ppedrot** [comment]
My take on that is that coinductive types should not enjoy dependent elimination, because in most models you can actually show that this does not hold. There is a very good reason for that, which is that cofixpoints have no canonical forms, even if you try performing η-expansions. For instance, how do you expand a cofixpoint that's doing two steps at once.

So, I think we should simply forbid coinductive types defined through constructors, and force people to use primitive records instead, as these elegantly solve the problem.

I was advocating for a warning on positively defined co-inductive types in 8.8, but I don't know if the discussion converged.

**Comment by maximedenes** [comment]
> I was advocating for a warning on positively defined co-inductive types in 8.8, but I don't know if the discussion converged.

Maybe a good topic for a WG discussion?

**Comment by herbelin** [comment]
@ppedrot: I know this view of you, but the current situation is that Coq's coinductive are actually greatest fixpoints of positive types (i.e. if I understand category theory language well enough, infinite limits of finite collmits) and not greatest fixpoints of negative types. 

And if I'm supporting to have negative types and infinite limits of negative types, as you are also supportive, I don't see why we should forbid to have infinite limits of positive types (as well actually as infinite colimits of negative types).

> So, I think we should simply forbid coinductive types defined through constructors, and force people to use primitive records instead, as these elegantly solve the problem.

I'm honestly not an expert in coinduction but I don't see why we should forbid to state and prove properties of streams. How would you do it without dependent elimination? How would you rephrase the above example with only negative types? Do you mean that you would like to only allow statements of the form `get infty = get (C infty)` (for `get` the name of the primitive projection)?

> There is a very good reason for that, which is that cofixpoints have no canonical forms, even if you try performing η-expansions. For instance, how do you expand a cofixpoint that's doing two steps at once.

Cofixpoints don't have less nor more a normal form than fixpoints. In which sense the question of normal forms would be different here than for fixpoints, e.g. what is the normal form of `Nat.add`? There is no way to seize arbitrary complex infinity by finite means, so the normal form of an infinite object can only be its (virtual) infinite unfolding in the metalanguage (i.e. the infinite stream `C (C (C ...))` in your question).

> I was advocating for a warning on positively defined co-inductive types in 8.8, but I don't know if the discussion converged.

Indeed, there are a lot to say about negative types and to the meaning of co-inductive types, and to the status of primitive projections (I actually even saw #6764 in the news...). I'm incidentally learning a lot by discussing with you, so let's continue.

**Comment by mattam82** [comment]
@herbelin you would change the [=] into [bisimilar] and use bisimilarity instead, as actually all libraries using coinductive types do AFAIK. Of course that begs the question, if we ever go to OTT/HoTT like equality, can we allow dependent elimination again, knowing that [=_Stream] is defined as bisimilarity in that case?

**Comment by herbelin** [comment]
> you would change the [=] into [bisimilar] and use bisimilarity instead, as actually all libraries using coinductive types do AFAIK.

Are you saying that in practice, nobody really uses dependent elimination of (Coq's positive presentation of) coinductive types and that the subject reduction failure is a useless artefact?

> if we ever go to OTT/HoTT like equality, can we allow dependent elimination again?

If my understanding is correct (and I assume that @ppedrot, you and me agree here), (finite) positive types do have dependent elimination and (finite) negative types don't. Actually, this is just by definition since negative types are defined by destructors, so they don't have an associated `match`.

So, if my understanding is correct (and again, I'm wondering if this coincides with your own view), the issue is with least and greatest fixpoints (say `μX.A` and `νX.A`). How to consider them? Should they be considered as positive and negative types respectively, or should they be considered as living in another conceptual dimension, i.e. neither as positive or negative (i.e. not defined from either constructors or destructors) but as finite notations for infinite expansions, with the only difference between them being that `μ` accepts infinity in the evaluation context side only and `ν` accepts infinity in the term side only? And then, under this view, the question of dependent elimination is irrelevant for `μX.A` and `νX.A`: it is relevant only for the underlying `A` which, it, shall be either positive or negative.

As a first step, maybe should we stop to use the "inductive" and "coinductive" confusing terminology and rather speak explicitly about types which are either finite negative, finite positive, infinite negative, infinite positive, coinfinite negative and coinfinite positive?

**Comment by mattam82** [comment]
I think if people really relied on dependent elimination of coinductives they would hit the subject reduction problem more often indeed (but that's just a guess). Once you've defined "s = cons (hd s) (tl s)" and tried to rewrite with it in a cofix proof and hit the productivity checker complaining, the main workaround is to switch to bisimulation relations, no?

**Comment by herbelin** [comment]
> Once you've defined "s = cons (hd s) (tl s)" and tried to rewrite with it in a cofix proof and hit the productivity checker complaining, the main workaround is to switch to bisimulation relations, no?

I have very little experience myself but I'm willing to believe this indeed.

**Comment by mattam82** [comment]
I'd like to summon @damien-pous in this discussion

**Comment by ppedrot** [comment]
> Cofixpoints don't have less nor more a normal form than fixpoints. In which sense the question of normal forms would be different here than for fixpoints, e.g. what is the normal form of Nat.add?

@herbelin That's precisely the point: **neither** cofixpoints **nor** fixpoints have a normal form.

Unluckily, cofixpoints *build* values, while fixpoints *destruct* values. So you should not be able to reason on the shape of a value built out of a cofixpoint. But the problem is that dependent elimination lets you do that for coinductive types, whereas you never reason on fixpoints for inductive types, instead reasoning on the properties of constructors.

If anything, we should have a form of dependent elimination for coinductive projections, i.e. for contexts accepting a coinductive type. I have no idea what the principle would look like though.

**Comment by jashug** [comment]
Wouldn't allowing cofix-expansion only on one side lead to a definitional equality which is either not symmetric (if expansion can only happen on the left) (staying inside the same universe level) or not transitive (if expansion can only happen on either the left or the right)?

**Comment by ppedrot** [comment]
@herbelin Also, I don't see how asymmetry would help in solving the following failure of SR:
```
CoInductive I := C : I -> I.
CoFixpoint infty := C infty.

Definition unfold : infty = infty :=
  match infty as x return match x with C n => x = x end with
  | C n => eq_refl (C n)
  end.
Fail Definition nf_unfold : infty = infty := Eval lazy in unfold.
```
If you want η-rules on cofixpoints, you quickly hit decidability issues. Better stick to ETT for that!

**Comment by jashug** [comment]
At least tangentially related is Conor McBride's http://strictlypositive.org/Ripley.pdf on not-quite decidable eta rules. This is what inspired my earlier comment.

**Comment by herbelin** [comment]
@ppedrot: the example you give is basically the same as the one I gave. The new conversion problem is `C infty = C infty ≤ infty = infty` which reduces to twice `C infty ≤ infty`, with asymmetry still in the form "type of subject" ≤ "object". The question of subject reduction is special and is intimately related to the question of expansion postponement.

The problem of [expansion postponement](http://tlca.di.unito.it/opltlca/opltlcasu30.html) was originally formulated in Pure Type Systems but it can probably be adapted to systems with arbitrary positive and negative types, and equality, and subtyping. Given a derivation, can we use conversion in the direction of expansion only once at the very end of the derivation? It has been proved by Barthe-Hatcliff-Sörensen that this holds for normalizing systems but the question for non-normalizing ones is still open. So, up to the question of ι, subtyping, ... which we should look at closely, it holds for Coq under the assumption of its normalization.

[Digression 1:  The [expansion postponement page](http://tlca.di.unito.it/opltlca/opltlcasu30.html) is however dubious. In my memory, Poll's proof was wrong and the Barthe-Hatcliff-Sörensen paper is not cited.]

[Digression 2: expansion postponement holds for injective systems, that have all dependent products (i.e. such that for any two sorts there is a sort such that the dependent product of types in the two first sorts is valid), even in the absence of a result of normalization. The reason expansion postponement holds in this case is that the only difficulty to postpone expansion is that subject reduction is needed in this premisse of the abstraction rule which justifies the validity of the product (so nothing to do if this premisse can be reconstructed canonically). But in subject reduction, to type the contractum of a β-redex, one needs to turn reductions in the type of the variable into expansions in the type of the instance of the variable at the position they are substituted. These new expansions have again to be propagated upwards, producing new subject reduction problems and noone found a criterion to break the loop yet. I don't know if this helps for Coq though, because we shall probably fall on the same situation for typing the return type of a "match", so to get expansion postponement for Coq, one would have to rely on its normalization.]

So, on the basis of the normalization of Coq and on the basis that Barthe-Hatcliff-Sörensen scales to Coq, subject reduction holds by needing conversion only in the direction of expansion.

This should also address the question from @jashug (Conor's Ripley paper is very nice by the way.). In particular, the situation of Conor is different. He is asking whether some term of type `A` can be given some type `B` and indeed, this may require reduction on both sides and thus symmetry.

Not saying that I have a proof that my "conjecture" holds but saying that the situation of proving subject reduction requires a constrained form of conversion.

@ppedrot:
> That's precisely the point: neither cofixpoints nor fixpoints have a normal form.
> If anything, we should have a form of dependent elimination for coinductive projections, i.e. for contexts accepting a coinductive type. I have no idea what the principle would look like though.

I think there is more to say here: e.g. if they don't have a normal form strictly speaking, can't we still consider "fueled" approximations of these normal forms, or do I miss something in the presence e.g. of side effects? But we shall probably have other opportunities to chat later.

**Comment by jashug** [comment]
It feels very strange to me to have
`eq_refl (C infty) : infty = C infty` succeed but
`eq_refl infty : infty = C infty` fail.
But as long as that is acceptable, I don't see an immediate solution of the halting problem.
(Note for future readers:
`C infty = C infty ≤ infty = infty` means `(@eq I (C infty) (C infty)) ≤ (@eq I infty infty)`,
not `(C infty) = (C infty) ≤ infty = infty`, this is a single inequality not a string of 3 inequalities)

**Comment by herbelin** [comment]
> It feels very strange to me to have `eq_refl (C infty) : infty = C infty` succeed but `eq_refl infty : infty = C infty` fail.

I agree that this asymmetry is strange but after all, conversion in Coq is already asymmetric (since it is actually a subtyping relation).

By the way, I'm more and more thinking that we should explore alternative presentations of typed equality which are not defined as the reflexive-symmetric-transitive-congruent closure of reduction, but in more operational ways. This is already what we do in Coq (where definitional equality is an ad hoc conversion algorithm based on delta-lazy weak-head normalization of both sides) but in an untyped way and the drawback is that this makes a long distance to the set-theoretic model (and in particular, we [lose](https://github.com/coq/coq/blob/8dd6d091ffbfa237f7266eeca60187263a9b521f/kernel/reduction.ml#L496-L499) the ability to interpret contravariant subtyping and full η). Or, alternatively, we should investigate fully typed presentation of the syntax-directed conversion algorithm of Coq (which certainly would have to look like Adam's Typed Parallel One-Step Reduction, TPOSR, so that confluence can be proved). [Incidentally, do you know people who would be interested to explore this kind of directions?]

**Comment by SkySkimmer** [comment]
>It feels very strange to me to have `eq_refl (C infty) : infty = C infty` succeed but `eq_refl infty : infty = C infty` fail.

It would mean `transport (fun i => infty = i) (eq_refl (C infty)) (eq_refl infty) : infty = C infty` but that reduces to `eq_refl infty : infty = C infty` so SR is still broken.

**Comment by mattam82** [comment]
I really doubt asymmetry is acceptable. Same kind of problem holds for making Acc definitionally proof-irrelevant. I think the more principled approach for corecursion is given by Guarded Cubical Type Theory where unfolding is controlled by propositional equalities and clocks.

**Comment by herbelin** [comment]
Nice counter-example, thanks. Even though it is still conversion in the sense of expansion which we need, this proof succeeds in inverting the effect of a demanding `match` in the "subject" towards the "object". So, yes, asymmetry is not enough. So, several possible conclusions:
- the subject reduction loss is anyway an artefact useless in practise; it is not worth exploring ways to solve it while preserving dependent elimination of greatest fixpoints (is that really the case? and what were the models @ppedrot was talking about in which dependent elimination of greatest fixpoints does not hold?),
- exploring whether a symmetric strategy may work, accepting unfolding on either side, but only on one side, once the side is chosen (@skyskimmer, even if only for fun and/or to get a more intimate knowledge of the fine-grain properties of conversion, do you see a possible counterexample if we do that?).

**Comment by jashug** [comment]
As I mentioned above, expanding on the left xor right gives a non-transitive relation.

**Comment by jashug** [comment]
Given `a : S -> option S`

Construct `a' : option S -> option S` by lifting
Gives a coalgebra for `F X = X`
Cofixpoint `sim (k : option S) = C (sim (a' k))`

Then if `i : S` halts, we have
`sim (Some i) ≤ C C .. C (sim None)` (halts in n steps)
and `C C .. C (sim None) ≤ (sim None)`
but `sim (Some i) ≤ (sim None)`

Therefore, `≤` is not transitive if we just expand on one side. (And is running up against the Halting Problem).

So, to break SR, we can have for example
`refl (sim (Some i)) : sim (Some i) = sim (Some i)`,
`id1 : (sim (Some i) = sim (Some i)) ->(C C .. C (sim None) = C C .. C (sim None)) := id1 x = x`
`id2 : (C C .. C (sim None) = C C .. C (sim None)) -> (sim None = sim None) := id2 x = x`
Then `id2 (id1 (refl (sim (Some i)))) : sim None = sim None` but reduces to
`refl (sim (Some i)) : sim None = sim None` which breaks SR.

**Comment by herbelin** [comment]
@jashug: Well, let me then continue to elaborate on the idea to separate expansion and reduction, or, actually, to just have two conversion rules. One is the decidable conversion we know attached in a syntax-directed way to typing rules (e.g. from `Γ ⊢ M : C`, `Γ ⊢ N : A`, `Γ ⊢ C →wh Πx:A'.B`, `Γ ⊢ A ≤ A'` get `Γ ⊢ MN : B[x:=N]`). The other is not requested to be decidable and is only used to "repair" subject reduction. The answer to `eq_refl infty :? infty = C infty` is no, i.e. "no, it does not hold for the dedicable conversion algorithm I'm using", but the ability of the metatheory to mechanically transform a derivation of `unfold : infty = C infty` into a derivation of `eq_refl infty : infty = C infty` is there, using an extended fragment of definitional equality which is not decidable (but still manipulable as soon as at some time we told the system that `unfold` contains enough information to reconstruct the part which the system is not able to construct alone). We could somehow compare such an approach to [Andromeda](http://www.andromeda-prover.org/) (preventively saying that I know that some opiniated people around do not like), which we can mention as an example of system where the difference between derivation and proof-term is emphasized.

Coming back on the other side to @ppedrot's proposal of restricting dependent positive elimination `match`, wouldn't it force to necessarily attach greatest fixpoint (ν-types) to the underlying positive structure? Accordingly, it is the case in Coq that ν-types are attached to the underlying structure, but if we think at ν-types as autonomous (as in Conor's Ripley paper), how could we restrict a `match` which does not make sense for ν-types properly speaking?

Told otherwise, don't we finally have only two solutions to the subject reduction problem:
- either attach ν-types to the underlying (necessarily non-trivial) structure so as to be able, in the positive case, to restrict `match` on the underlying positive structure,
- or "fix" the metatheory one way or another?

**Comment by herbelin** [comment]
Update: Pursuing my investigations, I arrive indeed to the conclusion of @ppedrot. There are two models of pure recursion/corecursion and in particular two models of `νX.A`:
- The, say "à la Church" (don't know if the expression is optimal), model with an explicit destructor `out` to control the unfolding of `ν` (at the type level) and of `cofix` (at the term level), as in Conor's Ripley paper (up to the interdefinability of `cofix` and of the pair `in`/`coit`). This model , by construction, does not support dependent elimination.
- The, say "à la Curry", model has no explicit destructor and the `ν` unfolds freely at the type level. This is the implicit model in current Coq. E.g. `Build_stream` is the positive constructor of `stream`, which comes with a dependent `match` without being surrounded by some `out`. As for `cofix`, it is only a finite notation for an infinite expansion, not a constructor itself. The counterpart is the difficulty in how to control the unfolding of the `cofix` by finite means as discussed in this issue.

**Comment by herbelin** [comment]
I add here for the sake of completeness the example of loss of subject reduction when defining a `fix` in a negative type equipped with η (dual then to the loss of subject reduction when defining a `cofix` in a positive type equipped with η, knowing that η in positive types is precisely expansion of the surrounding context). The example is a mix of @mattam82 example in 4341f37 and of @skyskimmer in the discussion about #407. 
```
Set Primitive Projections.
Record foo := C { p : foo }.
Definition f := fix f (x : foo) : nat := p x.

(* eta rule for foo *)
Definition tr (P : foo -> Type) (x : foo) (v : P (C (p x))) : P x := v.
Definition goal := forall x, f x = p x.

(* when we compute the tr away typechecking will fail *)
Definition thing : goal := fun x => (tr (fun x => f x = p x) x (eq_refl _)).
Definition thing_refl := eq_refl thing.
Fail Definition thing_refl' := Eval compute in thing_refl.
```
My analysis is that the conclusions to which I arrived in the last two comments applies as well in this case.
